discretenet/problems/gisp/gisp.py

```diff
@@ -0,0 +1,216 @@
+from pathlib import Path
+import random
+from typing import Union
+import pyomo.environ as pyo
+import networkx as nx
+try:
+    from importlib.resources import open_text
+except ImportError:
+    from importlib_resources import open_text
+
+from discretenet.problem import Problem
+from discretenet.generator import Generator
+
+
+class GISPProblem(Problem):
+    def __init__(self, graph, E2, name):
+        """
+        Generate a concrete Pyomo model of the GISP problem
+        """
+        self.model = None
+        self.name = name
+        self.graph = graph
+        self.E2 = list(E2)
+
+        # create concrete pyomo model
+        self.create_model()
+
+    def create_model(self):
+        model = pyo.ConcreteModel()
+        #  initialize variables and cost functions
+        coeffs = {}
+        model.nodes = pyo.Set(initialize=list(self.graph.nodes()))
+        model.x = pyo.Var(model.nodes, domain=pyo.Binary)
+        for node in self.graph.nodes():
+            # model.x[node] = pyo.Var(domain=pyo.Binary)
+            coeffs[node] = self.graph.nodes[node]["revenue"]
+
+        model.E2 = pyo.Set(initialize=self.E2)
+        model.y = pyo.Var(model.E2, domain=pyo.Binary)
+        for node1, node2 in self.E2:
+            # model.y[node1, node2] = pyo.Var(domain=pyo.Binary)
+            coeffs[node1, node2] = -1 * self.graph[node1][node2]["cost"]
+
+        #  objective function
+        model.objective = pyo.Objective(
+            expr=pyo.quicksum(
+                coeffs[node] * model.x[node] for node in self.graph.nodes()
+            )
+            + pyo.quicksum(
+                coeffs[node1, node2] * model.y[node1, node2] for node1, node2 in self.E2
+            ),
+            sense=pyo.maximize,
+        )
+
+        #  add constraints
+        model.constraints = pyo.ConstraintList()
+        for node1, node2, edge in self.graph.edges(data=True):
+            if (node1, node2) not in self.E2:
+                model.constraints.add(model.x[node1] + model.x[node2] <= 1)
+            else:
+                model.constraints.add(
+                    model.x[node1] + model.x[node2] - model.y[node1, node2] <= 1
+                )
+
+        self.model = model
+
+    def is_linear(self) -> bool:
+        return True
+
+    def get_name(self) -> str:
+        return self.name
+
+
+class GISPGenerator(Generator):
```

Comment:
Can you change this to `class GISPGenerator(Generator[GISPProblem])`? It's a generic type base class I setup recently to help with type hints for return values

Reply:
Updated